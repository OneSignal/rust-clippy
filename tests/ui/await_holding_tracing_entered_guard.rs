#![warn(clippy::await_holding_tracing_entered_guard)]

use tracing::{info_span, Span};

async fn bad(span: &Span) -> u32 {
    let _guard = span.enter();
    baz().await
}

async fn bad_mut(span: &Span) -> u32 {
    let _guard = span.enter();
    baz().await
}

async fn good(span: &Span) -> u32 {
    {
        let _guard = span.enter();
        let y = 1;
    }
    baz().await;
    let _guard = span.enter();
    47
}

async fn baz() -> u32 {
    42
}

async fn also_bad(span: &Span) -> u32 {
    let first = baz().await;

    let _guard = span.enter();

    let second = baz().await;

    let third = baz().await;

    first + second + third
}

async fn less_bad(span: &Span) -> u32 {
    let first = baz().await;

    let _guard = span.enter();

    let second = baz().await;

    drop(_guard);

    let third = baz().await;

    first + second + third
}

async fn not_good(span: &Span) -> u32 {
    let first = baz().await;

    let second = {
        let _guard = span.enter();
        baz().await
    };

    let third = baz().await;

    first + second + third
}

#[allow(clippy::manual_async_fn)]
fn block_bad(span: &Span) -> impl std::future::Future<Output = u32> + '_ {
    async move {
        let _guard = span.enter();
        baz().await
    }
}

fn main() {
    let span = info_span!("");
    good(&span);
    bad(&span);
    bad_mut(&span);
    also_bad(&span);
    less_bad(&span);
    not_good(&span);
    block_bad(&span);
}
