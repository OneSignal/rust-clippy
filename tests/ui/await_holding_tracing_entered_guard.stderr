error: this `Entered` held across an `await` point. To watch the runtime of a future, use `.instrument(span)`
  --> $DIR/await_holding_tracing_entered_guard.rs:6:9
   |
LL |     let _guard = span.enter();
   |         ^^^^^^
   |
   = note: `-D clippy::await-holding-tracing-entered-guard` implied by `-D warnings`
   = help: if you intend to create a span around a synchronous operation
note: these are all the `await` points this reference is held through
  --> $DIR/await_holding_tracing_entered_guard.rs:6:5
   |
LL | /     let _guard = span.enter();
LL | |     baz().await
LL | | }
   | |_^

error: this `Entered` held across an `await` point. To watch the runtime of a future, use `.instrument(span)`
  --> $DIR/await_holding_tracing_entered_guard.rs:11:9
   |
LL |     let _guard = span.enter();
   |         ^^^^^^
   |
   = help: if you intend to create a span around a synchronous operation
note: these are all the `await` points this reference is held through
  --> $DIR/await_holding_tracing_entered_guard.rs:11:5
   |
LL | /     let _guard = span.enter();
LL | |     baz().await
LL | | }
   | |_^

error: this `Entered` held across an `await` point. To watch the runtime of a future, use `.instrument(span)`
  --> $DIR/await_holding_tracing_entered_guard.rs:32:9
   |
LL |     let _guard = span.enter();
   |         ^^^^^^
   |
   = help: if you intend to create a span around a synchronous operation
note: these are all the `await` points this reference is held through
  --> $DIR/await_holding_tracing_entered_guard.rs:32:5
   |
LL | /     let _guard = span.enter();
LL | |
LL | |     let second = baz().await;
LL | |
...  |
LL | |     first + second + third
LL | | }
   | |_^

error: this `Entered` held across an `await` point. To watch the runtime of a future, use `.instrument(span)`
  --> $DIR/await_holding_tracing_entered_guard.rs:44:9
   |
LL |     let _guard = span.enter();
   |         ^^^^^^
   |
   = help: if you intend to create a span around a synchronous operation
note: these are all the `await` points this reference is held through
  --> $DIR/await_holding_tracing_entered_guard.rs:44:5
   |
LL | /     let _guard = span.enter();
LL | |
LL | |     let second = baz().await;
LL | |
...  |
LL | |     first + second + third
LL | | }
   | |_^

error: this `Entered` held across an `await` point. To watch the runtime of a future, use `.instrument(span)`
  --> $DIR/await_holding_tracing_entered_guard.rs:59:13
   |
LL |         let _guard = span.enter();
   |             ^^^^^^
   |
   = help: if you intend to create a span around a synchronous operation
note: these are all the `await` points this reference is held through
  --> $DIR/await_holding_tracing_entered_guard.rs:59:9
   |
LL | /         let _guard = span.enter();
LL | |         baz().await
LL | |     };
   | |_____^

error: this `Entered` held across an `await` point. To watch the runtime of a future, use `.instrument(span)`
  --> $DIR/await_holding_tracing_entered_guard.rs:71:13
   |
LL |         let _guard = span.enter();
   |             ^^^^^^
   |
   = help: if you intend to create a span around a synchronous operation
note: these are all the `await` points this reference is held through
  --> $DIR/await_holding_tracing_entered_guard.rs:71:9
   |
LL | /         let _guard = span.enter();
LL | |         baz().await
LL | |     }
   | |_____^

error: aborting due to 6 previous errors

